src/retry_logic.rs

```diff
@@ -38,15 +39,29 @@ impl RetryPolicyExt for RetryPolicy {
                 .maximum_interval
                 .clone()
                 .try_into_or_none()
-                .unwrap_or_else(|| interval * 100);
-            Some((interval.mul_f64(coeff.powi(attempt_number as i32 - 1))).min(max_iv))
+                .unwrap_or_else(|| interval.saturating_mul(100));
+            let mul_factor = coeff.powi(attempt_number as i32 - 1);
+            let tried_mul = try_from_secs_f64(mul_factor * interval.as_secs_f64());
+            Some(tried_mul.unwrap_or(max_iv).min(max_iv))
         } else {
             // No retries if initial interval is not specified
             None
         }
     }
 }
 
+const NANOS_PER_SEC: u32 = 1_000_000_000;
+/// modified from rust stdlib since this feature is currently nightly only
+fn try_from_secs_f64(secs: f64) -> Option<Duration> {
+    const MAX_NANOS_F64: f64 = ((u64::MAX as u128 + 1) * (NANOS_PER_SEC as u128)) as f64;
```

Comment:
Heh, I just saw that `9223372036854775807000000000` can't be represented as f64 but `9223372036854775808000000000` can.

General question, why can't we always operate in nanos and just do integer math? Go uses i64 as a duration and says "+/- 290 years is good enough".

Reply:
Just because then it'd be harder to use `Duration`s everywhere which are extremely convenient

Reply:
I think you could use `Duration::from_nanos` and `Duration::as_nanos` when you need to go from integer scalar to/from `Duration` (unless I'm misunderstanding).

Reply:
Oh, you can - but at least for this part the multiplication factor is in f64 and all this hoop jumping is to ensure the nanos portion *isn't* screwed up - it's actually just copied from the nightly stdlib for the most part (with the real error type stripped out), so I trust this method makes sense.

Reply:
K. I just wince at floats as a reflect. I prefer int math, but obv doesn't matter much here as whatever precision may be lost is minimal regardless of primitive chosen.